Enable rootfs ab redundancy support

dwalkes opened the pull request:
Add BSP support for enabling A/B redundancy and managing redundant and non-redundant A/B partition layouts, using the Nvidia provided BSP layouts.

When combined with https://github.com/OE4T/meta-tegra/pull/1285 , this logic provides a base for supporting A/B slot switching using uefi variables and capsule updates which is independent of a specific update tool used to write parititions.

Control for supporting A/B rootfs partition layout is supported through `USE_REDUNDANT_FLASH_LAYOUT` which is set to 1 by default for all platforms other than orin-nano (not enabling for orin-nano by default based on previous issues/observations noticed with https://github.com/OE4T/meta-tegra/discussions/1286).

Custom MACHINEs which already define their own `ROOTFSPART_SIZE`, `PARTITION_LAYOUT_TEMPLATE`, and `ROOTFSPART_SIZE` shouldn't be impacted by this change, although anyone who has done their own overlay implementation for Rootfs A/B may want to remove it in favor of this implementation.  Alternatively they can set `USE_REDUNDANT_FLASH_LAYOUT = 0` for their machine to handle on their own.

I've initially tested this on `jetson-xavier-nx-devkit-emmc` with this sequence:

In local.conf:
```
IMAGE_INSTALL:append = " tegra-uefi-capsules"
```

After building/booting, I see slots enabled and slot A selected

```
root@jetson-xavier-nx-devkit-emmc:~# nvbootctrl dump-slots-info
Current version: 35.4.1
Capsule update status: 0
Current bootloader slot: A
Active bootloader slot: A
num_slots: 2
slot: 0,             status: normal
slot: 1,             status: normal
```

After performing capsule update:
```
mkdir -p /opt/nvidia/esp/EFI/UpdateCapsule
cp /opt/nvidia/UpdateCapsule/tegra-bl.cap /opt/nvidia/esp/EFI/UpdateCapsule/TEGRA_BL.Cap
oe4t-set-uefi-OSIndications
```

and rebooting, I see:

```
root@jetson-xavier-nx-devkit-emmc:~# nvbootctrl dump-slots-info
Current version: 35.4.1
Capsule update status: 1
Current bootloader slot: B
Active bootloader slot: B
num_slots: 2
slot: 0,             status: normal
slot: 1,             status: normal
```
And rootfs changed to partition 2:
```
root@jetson-xavier-nx-devkit-emmc:~# mount | grep mmcblk0p
/dev/mmcblk0p2 on / type ext4 (rw,relatime)
```

If/when this looks ready to merge I'll test on a few other platforms and try some non-redundant configs.

Closes https://github.com/OE4T/meta-tegra/issues/1389

madisongh commented on conf/machine/jetson-orin-nano-devkit-nvme.conf:
```diff
@@ -4,8 +4,11 @@
-PARTITION_LAYOUT_TEMPLATE ?= "flash_t234_qspi.xml"
-PARTITION_LAYOUT_EXTERNAL ?= "flash_l4t_external.xml"
+PARTITION_LAYOUT_TEMPLATE_DEFAULT ?= "flash_t234_qspi.xml"
+PARTITION_LAYOUT_TEMPLATE_DEFAULT_SUPPORTS_REDUNDANT ?= "1"
```
Wouldn't hurt to add a comment here to explain why this is being set.  I think I get it, but it's not necessarily intuitive.

dwalkes replied:
I put a small dissertation about this in https://github.com/OE4T/meta-tegra/pull/1428/commits/e0e04e85d4acd1c34b55620d749d3f5180b14ac7 but can move some of it here, or maybe in the weak definition of `PARTITION_LAYOUT_TEMPLATE_DEFAULT_SUPPORTS_REDUNDANT`

madisongh replied:
Wiki page would be OK too.  Documenting it somewhere other than the commit history is helpful.

dwalkes replied:
I've started a wiki page at https://github.com/OE4T/meta-tegra/wiki/Rendundant-Rootfs-A-B-Partition-Support

madisongh commented on conf/machine/include/tegra-common.inc:
```diff
@@ -104,3 +105,13 @@ TEGRA_FLASHVARS ?= "BPFDTB_FILE BPF_FILE BR_CMD_CONFIG DEVICEPROD_CONFIG DEVICE_
 
 TEGRA_UEFI_SIGNING_CLASS ?= "tegra-uefi-signing"
+
+USE_REDUNDANT_FLASH_LAYOUT ?= "1"
```
Hmm.  Do we want this to be the default for everyone?

dwalkes replied:
I probably wouldn't have been brave enough to try it but I did based on https://github.com/OE4T/meta-tegra/issues/1389 and the comment about setting as default.  It seems a bit scary but if you are tegraflashing and your rootfs is small enough to fit it seems like a low risk.  My guess is few people use the default flash layouts for production too, but perhaps I'm wrong about that.  Added more about this in https://github.com/OE4T/meta-tegra/pull/1428/commits/e0e04e85d4acd1c34b55620d749d3f5180b14ac7 as well.

dwalkes replied:
Good idea, I did that at https://github.com/OE4T/tegra-demo-distro/pull/284

madisongh reviewed (changes requested):
Thanks @dwalkes , overall this looks good.  See in-line for some specific questions and changes.

dwalkes commented on conf/machine/include/tegra-common.inc:
```diff
@@ -104,3 +105,13 @@ TEGRA_FLASHVARS ?= "BPFDTB_FILE BPF_FILE BR_CMD_CONFIG DEVICEPROD_CONFIG DEVICE_
+PARTITION_LAYOUT_EXTERNAL ?= "${@d.getVar('PARTITION_LAYOUT_EXTERNAL_REDUNDANT') if d.getVar('USE_REDUNDANT_FLASH_LAYOUT') else d.getVar('PARTITION_LAYOUT_EXTERNAL_DEFAULT')}"
+
+ROOTFSPART_SIZE_REDUNDANT ?= "${@int(d.getVar('ROOTFSPART_SIZE_DEFAULT')) // 2}"
+ROOTFSPART_SIZE ?= "${@d.getVar('ROOTFSPART_SIZE_REDUNDANT') if d.getVar('USE_REDUNDANT_FLASH_LAYOUT') else d.getVar('ROOTFSPART_SIZE_DEFAULT')}"
```
Thanks, I believe I've addressed this with the most recent update.

dwalkes commented on recipes-bsp/uefi/edk2-firmware-tegra_35.4.1.bb:
```diff
@@ -20,6 +20,8 @@ NVDISPLAY_INIT ?= "${NVDISPLAY_INIT_DEFAULT}"
 NVDISPLAY_INIT_DEPS = ""
 NVDISPLAY_INIT_DEPS:tegra194 = "nvdisp-init:do_deploy"
 
+SRC_URI += "${@'file://L4TConfiguration-RootfsRedundancyLevelABEnable.dtso' if d.getVar('USE_REDUNDANT_FLASH_LAYOUT') else ''}"
```
Thanks!  Fixed.

dwalkes commented on conf/machine/include/tegra-common.inc:
```diff
@@ -13,6 +13,7 @@ UBOOT_EXTLINUX_KERNEL_ARGS ?= "${KERNEL_ARGS}"
 UBOOT_EXTLINUX_INITRD ?= "${@'/boot/initrd' if d.getVar('INITRAMFS_IMAGE') != '' and d.getVar('INITRAMFS_IMAGE_BUNDLE') != '1' else ''}"
 
 TEGRA_BOOTCONTROL_OVERLAYS ?= "L4TConfiguration.dtbo"
+TEGRA_BOOTCONTROL_OVERLAYS += "${@'L4TConfiguration-RootfsRedundancyLevelABEnable.dtbo' if d.getVar('USE_REDUNDANT_FLASH_LAYOUT') else ''}"
```
Thanks! Fixed.

dwalkes commented:
I believe I've addressed all the comments here.  I've also re-tested on `jetson-xavier-nx-devkit-emmc` in the default configuration with `USE_REDUNDANT_FLASH_LAYOUT = "0"` and the branch of tegra-demo-distro at https://github.com/OE4T/tegra-demo-distro/pull/284 which sets `USE_REDUNDANT_FLASH_LAYOUT_DEFAULT = "1"`.

Tomorrow I'll try with  `jetson-xavier-nx-devkit-emmc` and NVMe boot as well as Orin Nano with/without NVMe.

dwalkes commented:
Tested with jetson-xavier-nx-devkit-emmc
* non-redundant, emmc
* rendundant, emmc
* non-redundant, nvme
* redundant, nvme

Tested with jetson-orin-nano-devkit
* non-redundant, sdcard
* redundant, sdcard

Tested with jetson-orin-nano-devkit-emmc
* redundant, nvme

In each of the redundant cases I ran capsule update and verified the slots changed accordingly.

I think this is ready to merge but let me know if you'd like me to try something else.

I'd like to backport this to kirkstone next as well, let me know if you'd like a PR for this.

madisongh commented:
Thanks @dwalkes !  Once we've got a passing build on master (with your PR merged there), I'll cherry-pick these to the other branches.

madisongh commented:
I've backported these changes to nanbield, mickledore, and kirkstone branches.

arrow53 commented:
@dwalkes @madisongh with my orin nano on kirkstone with the tegra-demo-distro I don't see the redundant flash layout as being 1
```
/yocto/tegra-demo-distro-nvme/build$ bitbake -e demo-image-egl | grep ^USE_REDUNDANT_FLASH_LAYOUT
USE_REDUNDANT_FLASH_LAYOUT="0"
USE_REDUNDANT_FLASH_LAYOUT_DEFAULT="0"
```
I have these in my local.conf FWIW
```
IMAGE_INSTALL:append = " tegra-uefi-capsules"

TNSPEC_BOOTDEV = "nvme0n1p1"
```

dwalkes commented:
@arrow53 I think https://github.com/OE4T/tegra-demo-distro/pull/284 was never backported to kirkstone.  @madisongh would you prefer not to change tegra-demo-distro branches to avoid swapping partition layouts on stable branches?  I should probably just clarify wording in https://github.com/OE4T/meta-tegra/wiki/Redundant-Rootfs-A-B-Partition-Support#setting-up-a-custom-machine in that case.

madisongh commented:
@dwalkes Yes, that's right - #284 would have been a big surprise to folks already using kirkstone if it had been backported.
